src/inc/til/env.h

```diff
@@ -0,0 +1,575 @@
+            {
+                const auto hr = wil_env::GetComputerNameW<string_type, initialBufferLength>(result);
+                RETURN_HR_IF(hr, FAILED(hr) && (hr != HRESULT_FROM_WIN32(ERROR_ENVVAR_NOT_FOUND)));
+                return S_OK;
+            }
+
+#ifdef WIL_ENABLE_EXCEPTIONS
+            /** Looks up the computer name and fails if it is not found. */
+            template<typename string_type = wil::unique_cotaskmem_string, size_t initialBufferLength = MAX_COMPUTERNAME_LENGTH + 1>
+            string_type GetComputerNameW()
+            {
+                string_type result;
+                THROW_IF_FAILED((wil_env::GetComputerNameW<string_type, initialBufferLength>(result)));
+                return result;
+            }
+
+            /** Looks up the computer name and returns null if it is not found. */
+            template<typename string_type = wil::unique_cotaskmem_string, size_t initialBufferLength = MAX_COMPUTERNAME_LENGTH + 1>
+            string_type TryGetComputerNameW()
+            {
+                string_type result;
+                THROW_IF_FAILED((wil_env::TryGetComputerNameW<string_type, initialBufferLength>(result)));
+                return result;
+            }
+#endif
+
+            /** Looks up a registry value from 'key' and fails if it is not found. */
+            template<typename string_type, size_t initialBufferLength = 256>
+            inline HRESULT RegQueryValueExW(HKEY key, PCWSTR valueName, string_type& result) WI_NOEXCEPT
+            {
+                return wil::AdaptFixedSizeToAllocatedResult<string_type, initialBufferLength>(result,
+                                                                                              [&](_Out_writes_(valueLength) PWSTR value, size_t valueLength, _Out_ size_t* valueLengthNeededWithNul) -> HRESULT {
+                                                                                                  auto length = gsl::narrow<DWORD>(valueLength * sizeof(wchar_t));
+                                                                                                  const auto status = ::RegQueryValueExW(key, valueName, 0, nullptr, reinterpret_cast<BYTE*>(value), &length);
+                                                                                                  // length will receive the number of bytes including trailing null byte. Convert to a number of wchar_t's.
+                                                                                                  // AdaptFixedSizeToAllocatedResult will then resize buffer to valueLengthNeededWithNull.
+                                                                                                  // We're rounding up to prevent infinite loops if the data isn't a REG_SZ and length isn't divisible by 2.
+                                                                                                  *valueLengthNeededWithNul = (length + sizeof(wchar_t) - 1) / sizeof(wchar_t);

```

Comment:
This function is a little bit cheeky since `RegQueryValueExW` returns binary data, but this assumes the caller is trying to get a wide string. I think it's kind of fine though...